OR expression pushdown

findepi opened the pull request:
This includes OR pushdown framework for JDBC connectors, implemented
initially for PostgreSQL.
Along with this, there comes support for comparison expressions,
effective literals, generic literals (used e.g. for BIGINT), to support
a minimal viable usage example for the OR pushdown.

Relates to https://github.com/trinodb/trino/issues/18

findepi commented:
Currently based on https://github.com/trinodb/trino/pull/11045

findepi commented:
Rebased after #11045 merged.

findepi commented:
CI `ci / test (plugin/trino-hive, test-failure-recovery)` - https://github.com/trinodb/trino/issues/11196 / https://github.com/trinodb/trino/issues/10631

assaf2 commented on core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:
```diff
@@ -62,7 +67,13 @@
+import static java.util.Locale.ENGLISH;
 import static java.util.Objects.requireNonNull;
 
 public final class ConnectorExpressionTranslator
```
Add tests to `TestConnectorExpressionTranslator`

hashhar commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteOr.java:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        return Optional.of(terms.stream()
+                .collect(Collectors.joining(") OR (", "((", "))")));
```
The double parenthesis to ensure precedence of operations?

hashhar commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:
```diff
@@ -715,6 +715,13 @@ public void testLikePredicatePushdown()
+    @Test
+    public void testOrPredicatePushdown()
+    {
+        assertThat(query("SELECT * FROM nation WHERE name = 'ALGERIA' OR regionkey = 4"))
```
- two `!=`
- mix of `=` and `!=`
- assert no pushdown in case of a IS NULL comparision
- assert correctness if someone writes `= NULL` or `!= NULL`

findepi replied:
added

losipiuk commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteComparison.java:
```diff
@@ -0,0 +1,118 @@
+                .with(type().equalTo(BOOLEAN))
+                .with(argumentCount().equalTo(2))
+                .with(argument(0).matching(expression().capturedAs(LEFT)))
+                // Currently, LIKE's pattern must be a varchar.
```
copy paste artifact

findepi commented on core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:
```diff
@@ -227,13 +238,75 @@ public SqlToConnectorExpressionTranslator(Session session, Map<NodeRef<Expressio
+        @Override
+        protected Optional<ConnectorExpression> visitCast(Cast node, Void context)
+        {
+            if (isEffectivelyLiteral(plannerContext, session, node)) {
```
Not sure about this. What's complex about a constant? The explicit literals don't have such a check.

findepi commented:
(just rebased)

findepi commented on core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:
```diff
@@ -227,13 +238,75 @@ public SqlToConnectorExpressionTranslator(Session session, Map<NodeRef<Expressio
+
+        private static String functionNameForComparisonOperator(ComparisonExpression.Operator operator)
+        {
+            return "$" + operator.name().toLowerCase(ENGLISH);
```
Agreed. Also, there should be constants for these function names.

findepi commented on core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:
```diff
@@ -227,13 +238,75 @@ public SqlToConnectorExpressionTranslator(Session session, Map<NodeRef<Expressio
         }
 
+        @Override
+        protected Optional<ConnectorExpression> visitGenericLiteral(GenericLiteral node, Void context)
```
added tests

findepi commented on plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteComparison.java:
```diff
@@ -0,0 +1,118 @@
+    public RewriteComparison(Set<ComparisonOperator> enabledOperators)
+    {
+        Set<String> functionNames = enabledOperators.stream()
+                .map(operator -> "$" + operator.name().toLowerCase(ENGLISH))
```
reworked this

findepi commented:
AC. Changes as fixups (inline) + new commit `Remove redundant argument being field`.

hashhar commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:
```diff
@@ -692,6 +692,16 @@ public void testCharTrailingSpace()
+        assertThat(query("SELECT * FROM nation WHERE nationkey != 3 OR regionkey = 4")).isFullyPushedDown();
+        assertThat(query("SELECT * FROM nation WHERE nationkey != 3 OR regionkey != 4")).isFullyPushedDown();
+        assertThat(query("SELECT * FROM nation WHERE name = 'ALGERIA' OR regionkey = 4")).isFullyPushedDown();
+        assertThat(query("SELECT * FROM nation WHERE name IS NULL OR regionkey = 4")).isNotFullyPushedDown(FilterNode.class); // TODO `name IS NULL` is not pushed down
```
unrelated question: how can i find how a particular operator is modelled? What place to start looking in code? I search for things like `IsOperator` and comparision operators but can't find it.

findepi replied:
`AstBuilder`

hashhar commented:
*reminder to squash fixups before merge*

findepi commented:
squashed, rebased, resolved conflicts and added `Simplify TestConnectorExpressionTranslator` commit to do a TODO comment i am adding in an earlier commit

hashhar reviewed (approved):
LGTM from me.

hashhar commented on plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:
```diff
@@ -692,6 +692,16 @@ public void testCharTrailingSpace()
+    {
+        assertThat(query("SELECT * FROM nation WHERE nationkey != 3 OR regionkey = 4")).isFullyPushedDown();
+        assertThat(query("SELECT * FROM nation WHERE nationkey != 3 OR regionkey != 4")).isFullyPushedDown();
+        assertThat(query("SELECT * FROM nation WHERE name = 'ALGERIA' OR regionkey = 4")).isFullyPushedDown();
```
Does the expression rewrite depend on how predicates are ordered/grouped. I see that if add another case like

```
        assertThat(query("SELECT * FROM nation WHERE (name = 'ALGERIA' AND regionkey != 10) OR (regionkey = 4 AND name != 'FOOBAR')")).isFullyPushedDown();
```

it fails and we get a ScanFilterNode.

cc: @wendigo

findepi replied:
That's unrelated to OR. That's because `!=` isn't subsumed:

```
trino:tpch> EXPLAIN SELECT * FROM nation WHERE name != 'FOOBAR' ;
                                                           Query Plan
--------------------------------------------------------------------------------------------------------------------------------
....

 Fragment 1 [SOURCE]
     Output layout: [nationkey, name, regionkey, comment]
     Output partitioning: SINGLE []
     ScanFilter[table = postgresql:tpch.nation tpch.nation, filterPredicate = ("name" <> CAST('FOOBAR' AS varchar(25)))]
         Layout: [nationkey:bigint, name:varchar(25), regionkey:bigint, comment:varchar(152)]
         Estimates: {rows: 25 (3.13kB), cpu: 3.13k, memory: 0B, network: 0B}/{rows: ? (?), cpu: 6.25k, memory: 0B, network: 0B}
         nationkey := nationkey:bigint:int8
         regionkey := regionkey:bigint:int8
         name := name:varchar(25):varchar
         comment := comment:varchar(152):varchar
```
